### PR TITLE
Add support for moveItemAtURL and unit tests.

### PR DESCRIPTION
**Pull request by msft-Jeyaram**
#653  #662

### CODE REVIEW

File: Frameworks/Foundation/NSFileManager.mm [hunk 1]
```diff
@@ -436,36 +436,48 @@ - (BOOL)copyItemAtPath:(id)srcPath toPath:(id)destPath error:(NSError**)error {
...
 /**
- @Status Stub
+ @Status Caveat
+ @Notes does not support file manager move delegates for shouldMoveItem or shouldProceedAfterError.
 */
 - (BOOL)moveItemAtURL:(NSURL*)srcURL toURL:(NSURL*)dstURL error:(NSError**)error {
-    UNIMPLEMENTED();
-
-    return NO;
+    return [self moveItemAtPath:[srcURL path] toPath:[dstURL path] error:error];
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSFileManager.mm]: This might do well to check that the file URLs are file URLs; it doesn't make sense to move anything else.

File: Frameworks/Foundation/NSFileManager.mm [hunk 2]
```diff
@@ -436,36 +436,48 @@ - (BOOL)copyItemAtPath:(id)srcPath toPath:(id)destPath error:(NSError**)error {
...
+            *error = [NSError errorWithDomain:NSCocoaErrorDomain code:NSFileNoSuchFileError userInfo:nil];
+        }
+
+        return NO;
+    }
+
     const char* src = [srcPath UTF8String];
     const char* dest = [destPath UTF8String];
 
     TraceVerbose(TAG, L"Moving %hs to %hs", src, dest);
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSFileManager.mm]: >TraceVerbose(TAG, L"Moving %hs to %hs", src, dest); [](start = 4, length = 51)

We can probably destroy this log.

File: tests/unittests/Foundation/NSFileManagerTests.mm [hunk 3]
```diff
@@ -136,4 +170,76 @@
...
+    EXPECT_OBJCEQ_MSG(str, content, "FAILED: Unable to read the file content.");
+
+    [fh closeFile];
+
+    deleteFile((status ? destName : srcName));
+}
+
+TEST(NSFileManager, MoveFileViaURL) {
+    char const* srcName = "\\NSFileManagerMoveTestFile.txt";
+    char const* destName = "\\MovedFile.txt";
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: > "\\ [](start = 26, length = 4)

The slashes here worry me; we should be using something like `-[NSString stringByAddingPathComponent:]` and its friends rather than relying on the input string to have a \ at the beginning.

File: tests/unittests/Foundation/NSFileManagerTests.mm [hunk 4]
```diff
@@ -14,18 +14,52 @@
...
+    int len = GetModuleFileNameA(NULL, fullPath, _MAX_PATH);
+    char* ptrToLastEntry = strrchr(fullPath, '\\');
+    int lengthOfLastEntry = strlen(ptrToLastEntry);
+    strncpy_s(ptrToLastEntry, _MAX_PATH - (len - lengthOfLastEntry), filePath, strlen(filePath) + 1);
+
+    return [NSString stringWithFormat:@"%s", fullPath];
+}
+
+static void createFileWithContentAndVerify(char const* filePath, NSString* content) {
+    NSString* fullPath = getPathToFile(filePath);
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: This entire function could be [content writeToFile:xxx atomically:NO];

File: tests/unittests/Foundation/NSFileManagerTests.mm [hunk 5]
```diff
@@ -14,18 +14,52 @@
...
+#import <Foundation/NSDate.h>
+#import <Foundation/NSURL.h>
+#import <Foundation/NSFileHandle.h>
+#import <stdlib.h>
+#import <windows.h>
+
+static NSString* getPathToFile(char const* filePath) {
+    char fullPath[_MAX_PATH];
+    int len = GetModuleFileNameA(NULL, fullPath, _MAX_PATH);
+    char* ptrToLastEntry = strrchr(fullPath, '\\');
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: pre-cache this, and simply use `[_cachedModulePath stringByAppendingPathComponent:filePathWhichIsNowAnNSString]`

**Comment by DHowett-MSFT** [comment]
:clock1:

**Comment by DHowett-MSFT** [comment]
If you explicitly say `Fixes #xxx.` or `Fix #xxx.` in the commit log, it will automatically close the issue when the pull request is merged!

File: Frameworks/Foundation/NSFileManager.mm [hunk 6]
```diff
@@ -302,18 +302,18 @@ - (BOOL)createDirectoryAtPath:(NSString*)pathAddr
                 bool success = EbrMkdir(curPath);
                 if (!success && errno != EEXIST) {
                     TraceError(TAG, L"Failed to make path %hs: %d", curPath, errno);
-                    // return FALSE;
+                    // return NO;
```
Comment (MSFTFox) [inline Frameworks/Foundation/NSFileManager.mm]: Is there a reason this is commented out and not just removed? Also, it seems like the error parameter is unused. This is probably outside the scope of this review, but could you take a look at it and file issues if necessary or follow up with somebody?

Comment (msft-Jeyaram) [inline Frameworks/Foundation/NSFileManager.mm]: there are allot of things that needs to be cleaned up in this file + most of the functionality does not have unit tests. I opened a separate bug for that.
So far this review is for algo's bug fix.

---
In reply to: [71405791](https://github.com/Microsoft/WinObjC/pull/669#discussion_r71405791) [](ancestors = 71405791)

File: Frameworks/Foundation/NSFileManager.mm (hunk 2 above)
Comment (MSFTFox) [inline Frameworks/Foundation/NSFileManager.mm]: In the wake of these logs, could you scrub this file for any unnecessary logging? It might also be worth looking into standardizing the error messages while you're in here since you seem to be touching on those a bit.

File: tests/unittests/Foundation/NSFileManagerTests.mm [hunk 7]
```diff
@@ -136,4 +170,76 @@
...
+    NSFileManager* manager = [NSFileManager defaultManager];
+
+    NSError* error = nil;
+    BOOL status = [manager moveItemAtPath:srcPath toPath:destPath error:&error];
+    EXPECT_TRUE(status);
+    EXPECT_EQ(nil, error);
+
+    // Verify file exists.
+    EXPECT_TRUE([manager fileExistsAtPath:destPath]);
+    EXPECT_FALSE([manager fileExistsAtPath:srcPath]);
```
Comment (MSFTFox) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: Is there a reason these are expects and not asserts? Just curious.

File: Frameworks/Foundation/NSFileManager.mm (hunk 2 above)
Comment (msft-Jeyaram) [inline Frameworks/Foundation/NSFileManager.mm]: Yep, the error reflects reference platform, except for rename call, reference platform sets type to POSIX, we are setting it to WIN32. although the code is the errno that is being set.

Overall file clean up/error standardization  will be part of #662

---
In reply to: [71406908](https://github.com/Microsoft/WinObjC/pull/669#discussion_r71406908) [](ancestors = 71406908)

File: tests/unittests/Foundation/NSFileManagerTests.mm [hunk 8]
```diff
@@ -136,4 +170,76 @@
...
+
+    NSString* srcPath = getPathToFile(srcName);
+    NSString* destPath = getPathToFile(destName);
+
+    NSFileManager* manager = [NSFileManager defaultManager];
+
+    NSError* error = nil;
+    BOOL status = [manager moveItemAtPath:srcPath toPath:destPath error:&error];
+    EXPECT_TRUE(status);
+    EXPECT_EQ(nil, error);
```
Comment (MSFTFox) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: This Expect nil, error really doesn't do a whole lot. It might be worthwhile to add a test for actual errors instead of a test that just makes sure nil comes back as nil through a function.

Comment (msft-Jeyaram) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: this is to make sure that although the file was moved, we don't have any error flags set.
Will add negative tests.

---
In reply to: [71407552](https://github.com/Microsoft/WinObjC/pull/669#discussion_r71407552) [](ancestors = 71407552)

**Comment by msft-Jeyaram** [comment]
somehow I lost that in my squash and ended up adding it manually.

---
In reply to: [233735940](https://github.com/Microsoft/WinObjC/pull/669#issuecomment-233735940) [](ancestors = 233735940)

File: tests/unittests/Foundation/NSFileManagerTests.mm [hunk 9]
```diff
@@ -136,4 +170,76 @@
...
+
+    NSData* fileData = [fh readDataToEndOfFile];
+    EXPECT_TRUE_MSG(fileData != nil, "FAILED: reading the entire file failed.");
+
+    NSString* str = [[[NSString alloc] initWithData:fileData encoding:NSUTF8StringEncoding] autorelease];
+    EXPECT_OBJCEQ_MSG(str, content, "FAILED: Unable to read the file content.");
+
+    [fh closeFile];
+
+    deleteFile((status ? destName : srcName));
```
Comment (MSFTFox) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: The deletes here worry me a bit. In a really bad situation, we could have files lying around. It might be best to move any deletes to a separate test so they get sweeped regularly instead of at the end of their own tests. Something like a test at the end of this file that checks if these files exist and deletes them if they do would be great.

Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: Prefer a WIL scope defer, actually. RAII is king!

---
In reply to: [71407993](https://github.com/Microsoft/WinObjC/pull/669#discussion_r71407993) [](ancestors = 71407993)

File: tests/unittests/Foundation/NSFileManagerTests.mm [hunk 10]
```diff
@@ -136,4 +170,76 @@
...
+    EXPECT_FALSE([manager fileExistsAtPath:[srcURL path]);
+
+    // Verify data.
+    NSFileHandle* fh = [NSFileHandle fileHandleForReadingAtPath:getPathToFile(destName)];
+    EXPECT_TRUE(fh != nil);
+
+    NSData* fileData = [fh readDataToEndOfFile];
+    EXPECT_TRUE_MSG(fileData != nil, "FAILED: reading the entire file failed.");
+
+    NSString* str = [[[NSString alloc] initWithData:fileData encoding:NSUTF8StringEncoding] autorelease];
```
Comment (MSFTFox) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: Might it be worthwhile to add a test for data other than UTF8?

Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: This is not testing whether we can read UTF-8, and additional encodings are out-of-scope here.
It'd probably be better to avoid using NSString entirely for this and just compare two NSData instances.

---
In reply to: [71408245](https://github.com/Microsoft/WinObjC/pull/669#discussion_r71408245) [](ancestors = 71408245)

File: tests/unittests/Foundation/NSFileManagerTests.mm (hunk 7 above)
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: `dest` should be an ASSERT, and `src` should be an EXPECT. The destination not existing will be fatal, but the source not existing will not impact the rest of the test.

---
In reply to: [71407123](https://github.com/Microsoft/WinObjC/pull/669#discussion_r71407123) [](ancestors = 71407123)

File: tests/unittests/Foundation/NSFileManagerTests.mm [hunk 11]
```diff
@@ -136,4 +170,76 @@
...
+
+    // Verify data.
+    NSFileHandle* fh = [NSFileHandle fileHandleForReadingAtPath:destPath];
+    EXPECT_TRUE(fh != nil);
+
+    NSData* fileData = [fh readDataToEndOfFile];
+    EXPECT_TRUE_MSG(fileData != nil, "FAILED: reading the entire file failed.");
+
+    NSString* str = [[[NSString alloc] initWithData:fileData encoding:NSUTF8StringEncoding] autorelease];
+    EXPECT_OBJCEQ_MSG(str, content, "FAILED: Unable to read the file content.");
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: It seems unusual to use NSFileHandle here; NSData is equally capable of reading a file from disk in a single line, and makes for a more succinct test surface.

File: tests/unittests/Foundation/NSFileManagerTests.mm (hunk 9 above)
Comment (msft-Jeyaram) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: wrapped up with RAII 

---
In reply to: [71407993](https://github.com/Microsoft/WinObjC/pull/669#discussion_r71407993) [](ancestors = 71407993)

File: tests/unittests/Foundation/NSFileManagerTests.mm (hunk 11 above)
Comment (msft-Jeyaram) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: indeed. switched to NSData, will makes it easier to verify data.

---
In reply to: [71408764](https://github.com/Microsoft/WinObjC/pull/669#discussion_r71408764) [](ancestors = 71408764)

File: tests/unittests/Foundation/NSFileManagerTests.mm (hunk 7 above)
Comment (msft-Jeyaram) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: ASSERT: Fails fast, aborting the current function.
    EXPECT: Continues after the failure.

Updating, also using RAII, that should make life easier :D

---
In reply to: [71408625](https://github.com/Microsoft/WinObjC/pull/669#discussion_r71408625) [](ancestors = 71408625,71407123)

File: Frameworks/Foundation/NSFileManager.mm (hunk 1 above)
Comment (msft-Jeyaram) [inline Frameworks/Foundation/NSFileManager.mm]: I was letting it fall down to the fact if it's not fileURL then we throw exception. looks like reference platform returns an error with the exception.

---
In reply to: [71392720](https://github.com/Microsoft/WinObjC/pull/669#discussion_r71392720) [](ancestors = 71392720)

File: Frameworks/Foundation/NSFileManager.mm [hunk 12]
```diff
@@ -1,8 +1,8 @@
 //******************************************************************************
 //
-// Copyright (c) 2015 Microsoft Corporation. All rights reserved.
+// COPYRIGHT (C) MICROSOFT. ALL RIGHTS RESERVED.
 //
-// This code is licensed under the MIT License (MIT).
+// THIS CODE IS LICENSED UNDER THE MIT LICENSE (MIT).
```
Comment (msft-Jeyaram) [inline Frameworks/Foundation/NSFileManager.mm]: oh god everything got caped.

File: tests/unittests/Foundation/NSFileManagerTests.mm [hunk 13]
```diff
@@ -136,4 +161,55 @@
...
+    EXPECT_EQ(nil, error);
+
+    // Verify file exists.
+    ASSERT_TRUE([manager fileExistsAtPath:[destURL path]]);
+    EXPECT_FALSE([manager fileExistsAtPath:[srcURL path]]);
+
+    // Verify data.
+
+    ASSERT_OBJCEQ([content dataUsingEncoding:NSUTF8StringEncoding], [NSData dataWithContentsOfURL:destURL]);
+}
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: It might be even better to sue `ASSERT_OBJCEQ(content, [NSString stringWithContentsOfFile:...])` to avoid encoding shenanigans.

Comment (msft-Jeyaram) [inline tests/unittests/Foundation/NSFileManagerTests.mm]: either way we end up encoding, but it's a single pass. changed .

---
In reply to: [71443437](https://github.com/Microsoft/WinObjC/pull/669#discussion_r71443437) [](ancestors = 71443437)

**Comment by rajsesh-msft** [comment]
:shipit:

**Comment by mnithish** [comment]
CI build successful
